### PR TITLE
Module implementation

### PR DESCRIPTION
**Pull request by sesh-kebab**
First commit - adds functionality for the terraform module including documentation (README.md) and development workflow (DEVELOP.md).

**How to test it**
1. Ensure you have a layer0 instance by creating a new one or reusing an existing one
2. Navigate to the `examples` directory and populate the variables either via a `tfvars` file like the one below:
```
endpoint = "https://<l0-endpoint>"
token = "<l0-token>"
access_key = "abc"
secret_key = "def"
region = "us-west-2"
kinesis_stream_name = "salesforce_splunk"
stream_aws_access_key = "<dev-ops-access-key>"
stream_aws_secret_key = "<dev-ops-secret-key>"
stream_aws_region = "us-west-2"
layer0_prefix = "<l0-prefix>"
``` 
3. Apply the terraform file
4. Create another service in the same layer0 instance which generates cloudwatch logs. You can use something like our guestbook sample which creates log entries for every entry you add to the guestbook.
5. Navigate to our splunk instance and query `host=layer0`. 
6. Confirm you can see the logs from your service in Splunk.

**Note:**
The example directory has a sample terraform configuration for how to use this module. Note however that the module points to a github repo which doesn't exist yet as this PR is intended to add the module to this repo.

So just for the test, in `examples/main.tf`

Instead of:
```
module "layer0_log_forwarder" {
  source                = "github.com/quintiles/layer0-log-forwarder/terraform"
  ...
}
```

Update it to a relative path:
```
module "layer0_log_forwarder" {
  source                = "./../terraform"
  ...
}
```

Also for the stream name - you can use `salesforce_splunk` in the devops account as no one is writing to that stream currently.

**Caveats**
There seems to be an issue with the aws terraform resrouce: `aws_cloudwatch_log_subscription_filter`. The state of this seems to be eventually consistent, which in my observations leads to creates, deletes and recreations to fail.

### CODE REVIEW

File: DEVELOP.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Layer0 Log Forwarder
+
+This terraform module forwards logs generated by services and tasks deployed via Layer0 to a Kinesis stream.
+
+## Minimum Requirements:
+
+* Terraform v0.11.3
+* Node.js v6.10
```
Comment (diemonster) [inline DEVELOP.md:8]: I'm just going to play devil's advocate here for a moment and ask: why node? Lambda supports Go now, is there a compelling reason to use Node aside from the code already being written that way?

File: terraform/src/event.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{ 
+  "awslogs": {
+    "data": "H4sIAAAAAAAAAIWQzWrDMBCEX8XoHONdraSVfAvUzam04NxKKLKjpqZxbGylJYS8e5WUnkrpaeHbn5mds+jDPPtdWJ/GIEpxt1wvXx6qul6uKrEQw+chTAlbbQiMcwqMTXg/7FbTcBxTZw/5HOa37pvWcQq+v+FidwxzbIbhvbDQGNP6be6lDrkKyue2xTZ3jWXndRs8qrQ/H5u5nboxdsPhvtvHMM2ifBahH+Mpe70BsbnJVB/hEK/Ns+i2SY3IOqWJpdNOWUmaHBqJ2hoFzOiskloDIaZBttaRBIcOrqKxSwFE36dfUKdJSwbZsV38BJPOS0BbABWAGehSU0mcPT3W66zIxGXxy0OKSoNiQsXOAhviVJVCBUoaJkapmUAyc4oV//Tg1H8eVtW3hc3lC2cxHuLHAQAA"
```
Comment (diemonster) [inline terraform/src/event.json:3]: What is in this `data` value?

File: terraform/src/index.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+const ecs = new aws.ECS();
+const kinesis = new aws.Kinesis({
+    accessKeyId: process.env.STREAM_AWS_ACCESS_KEY,
+    secretAccessKey: process.env.STREAM_AWS_SECRET_KEY,
+    region: process.env.STREAM_AWS_REGION
+});
+
+const l0Cache = new Layer0Cache(ecs);
+
+exports.handler = function(event, context, callback) {
```
Comment (diemonster) [inline terraform/src/index.js:15]: I don't feel like I have much to say regarding this code that hasn't already been said... it makes enough sense to me, flow-wise, but I'm not proficient enough with node to make very meaningful critiques in terms of style or best-practices.

File: terraform/src/index.js [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    if (streamNameParts.length == 3) {
+        taskGuid = streamNameParts[2];
+    }
+
+    return taskGuid;
+}
+
+// attempt to determine the severity based on the content
+function extractSeverity(logText) {
```
Comment (diemonster) [inline terraform/src/index.js:145]: Are these severity levels meant to target log output in v0.10.xv specifically?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+For Example:
+
+##### Cloudwatch log event
+```
+[Warning] Running out of disk space
+```
+
+##### What gets written to the Kinesis stream
+```json
```
Comment (diemonster) [inline README.md:19]: this structure is really nice 👍

**Review by diemonster — COMMENTED** [review]
some initial thoughts...

File: README.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+{
+    "Severity": "Warning",
+    "Message": "[WARNING] Running out of disk space",
```
Comment (andycmaj) [inline README.md:22]: how does this fit in with apps that already have an existing framework-specific format, ie ImsHealth.Logging?

File: DEVELOP.md (hunk 1 above)
Comment (sesh-kebab) [reply DEVELOP.md:8]: There is no compelling reason that I can think of. Would definitely look into creating a Go Lambda, but have not done so yet; so I can't make any comparisons.

I will say that I do like the fact that you just distribute the source for a node.js lambda where as for a go lambda, you would have to distribute the built binary as part of the terraform module.

File: terraform/src/event.json (hunk 2 above)
Comment (sesh-kebab) [reply terraform/src/event.json:3]: This is what a lambda that subscribes to cloudwatch events receives as an event. It is a base64 encoded, gzip'd json text.

When decoded and uncompressed, it is 2 sample events. Event.json is used for local development - so when we execute `node-lamda run`, the contents of this file are passed in to the handler which essentially simulates what would happen when you actually deploy the lambda and it receives a cloudwatch event.

File: terraform/src/index.js (hunk 4 above)
Comment (sesh-kebab) [reply terraform/src/index.js:145]: These are just arbitrary levels. Our current logging library (ImsHealth.Logging) has these levels so all the events that are logged from there have one of these severity levels set. This is more just being consistent with the other structured log data we are currently sending to splunk.

File: README.md (hunk 5 above)
Comment (sesh-kebab) [reply README.md:19]: 👍

File: README.md (hunk 6 above)
Comment (sesh-kebab) [reply README.md:22]: This structure mimics what is output by ImsHealth.Logging. So the intention is that the user of this module would forward to one of our logging pipeline streams and a lambda forwarder would be able to pick it up and forward to the hec endpoints.

If you search in splunk for `host=layer0` you will see some of the data that has been sent via this module.

**Pull request merged**